File: julia/src/data_structures/queue.jl
"""
Queue - FIFO data structure

Time Complexity:
- enqueue!: O(1) amortized
- dequeue!: O(1) amortized
- peek: O(1)
- isempty: O(1)
- size: O(1)
"""

mutable struct Queue{T}
    items::Vector{T}
    front::Int

    Queue{T}() where T = new{T}(T[], 1)
end

Queue() = Queue{Any}()

"""
    enqueue!(q::Queue{T}, item::T) where T

Add an item to the back of the queue.
"""
function enqueue!(q::Queue{T}, item::T) where T
    push!(q.items, item)
    q
end

"""
    dequeue!(q::Queue{T}) where T

Remove and return the front item.
Throws ArgumentError if queue is empty.
"""
function dequeue!(q::Queue{T}) where T
    if isempty(q)
        throw(ArgumentError("Queue is empty"))
    end
    result = q.items[q.front]
    q.front += 1

    # Compact if too much wasted space
    if q.front > length(q.items) ÷ 2 && q.front > 16
        q.items = q.items[q.front:end]
        q.front = 1
    end

    result
end

"""
    peek(q::Queue{T}) where T

Return the front item without removing it.
Throws ArgumentError if queue is empty.
"""
function peek(q::Queue{T}) where T
    if isempty(q)
        throw(ArgumentError("Queue is empty"))
    end
    q.items[q.front]
end

"""
    isempty(q::Queue)

Check if the queue is empty.
"""
Base.isempty(q::Queue) = q.front > length(q.items)

"""
    size(q::Queue)

Return the number of items in the queue.
"""
Base.size(q::Queue) = length(q.items) - q.front + 1

"""
    length(q::Queue)

Return the number of items in the queue.
"""
Base.length(q::Queue) = max(0, length(q.items) - q.front + 1)

"""
    empty!(q::Queue)

Remove all items from the queue.
"""
function Base.empty!(q::Queue)
    empty!(q.items)
    q.front = 1
    q
end

"""
    collect(q::Queue)

Convert queue to array (front to back).
"""
function Base.collect(q::Queue)
    if isempty(q)
        return eltype(q.items)[]
    end
    q.items[q.front:end]
end

File: lua/data_structures/queue.lua
-- Queue - FIFO data structure
--
-- Time Complexity:
-- - enqueue: O(1)
-- - dequeue: O(1) amortized
-- - peek: O(1)
-- - isEmpty: O(1)
-- - size: O(1)

local Queue = {}
Queue.__index = Queue

function Queue.new()
    local self = setmetatable({}, Queue)
    self.items = {}
    self.front = 1
    self.back = 0
    return self
end

function Queue:enqueue(item)
    self.back = self.back + 1
    self.items[self.back] = item
end

function Queue:dequeue()
    if self:isEmpty() then
        error("Queue is empty")
    end
    local item = self.items[self.front]
    self.items[self.front] = nil
    self.front = self.front + 1

    -- Compact if too much wasted space
    if self.front > 16 and self.front > self:size() then
        local newItems = {}
        local newIndex = 1
        for i = self.front, self.back do
            newItems[newIndex] = self.items[i]
            newIndex = newIndex + 1
        end
        self.items = newItems
        self.back = self.back - self.front + 1
        self.front = 1
    end

    return item
end

function Queue:peek()
    if self:isEmpty() then
        error("Queue is empty")
    end
    return self.items[self.front]
end

function Queue:isEmpty()
    return self.front > self.back
end

function Queue:size()
    return self.back - self.front + 1
end

function Queue:clear()
    self.items = {}
    self.front = 1
    self.back = 0
end

function Queue:toArray()
    local result = {}
    local index = 1
    for i = self.front, self.back do
        result[index] = self.items[i]
        index = index + 1
    end
    return result
end

return Queue
